tensorflow/core/kernels/generate_box_proposals_op.cu.cc

```diff
@@ -0,0 +1,594 @@
+    Tensor dev_sorted_scores;            // sorted scores, cub output
+    Tensor dev_boxes;                    // boxes on device
+    Tensor dev_boxes_keep_flags;  // bitmask for keeping the boxes or rejecting
+                                  // from output
+    const int nboxes_to_generate = std::min(conv_layer_nboxes, pre_nms_topn);
+    OP_REQUIRES_OK(
+        context,
+        AllocateGenerationTempTensors(
+            context, &d_conv_layer_indexes, &d_image_offset, &d_cub_sort_buffer,
+            &d_cub_select_buffer, &d_sorted_conv_layer_indexes,
+            &dev_sorted_scores, &dev_boxes, &dev_boxes_keep_flags, num_images,
+            conv_layer_nboxes, cub_sort_temp_storage_bytes,
+            cub_select_temp_storage_bytes, nboxes_to_generate, box_dim));
+    const GPUDevice& d = context->eigen_device<GPUDevice>();
+    Cuda2DLaunchConfig conf2d =
+        GetCuda2DLaunchConfig(conv_layer_nboxes, num_images, d);
+    // create box indices and offsets for each image on device
+    TF_CHECK_OK(GpuLaunchKernel(InitializeDataKernel, conf2d.block_count,
+                                conf2d.thread_per_block, 0, d.stream(), conf2d,
+                                d_image_offset.flat<int>().data(),
+                                d_conv_layer_indexes.flat<int>().data()));
+
+    // sort boxes with their scores.
+    // d_sorted_conv_layer_indexes will hold the pointers to old indices.
+
+    TF_OP_REQUIRES_CUDA_SUCCESS(
+        context,
+        cub::DeviceSegmentedRadixSort::SortPairsDescending(
+            d_cub_sort_buffer.flat<int8>().data(), cub_sort_temp_storage_bytes,
+            scores.flat<float>().data(), dev_sorted_scores.flat<float>().data(),
+            d_conv_layer_indexes.flat<int>().data(),
+            d_sorted_conv_layer_indexes.flat<int>().data(),
+            num_images * conv_layer_nboxes, num_images,
+            d_image_offset.flat<int>().data(),
+            d_image_offset.flat<int>().data() + 1, 0,
+            8 * sizeof(float),  // sort all bits
+            cuda_stream));
+    // Keeping only the topN pre_nms
+    conf2d = GetCuda2DLaunchConfig(nboxes_to_generate, num_images, d);
+
+    // create box y1,x1,y2,x2 from box_deltas and anchors (decode the boxes) and
+    // mark the boxes which are smaller that min_size ignored.
+    if (use_detectron_offset_) {
+      TF_CHECK_OK(GpuLaunchKernel(
+          GeneratePreNMSUprightBoxesKernel<true>, conf2d.block_count,
+          conf2d.thread_per_block, 0, d.stream(), conf2d,
+          d_sorted_conv_layer_indexes.flat<int>().data(),
+          reinterpret_cast<const float4*>(bbox_deltas.flat<float>().data()),
+          reinterpret_cast<const float4*>(anchors.flat<float>().data()), height,
+          width, num_anchors, min_size, image_info.flat<float>().data(),
+          bbox_xform_clip_default_,
+          reinterpret_cast<float4*>(dev_boxes.flat<float>().data()),
+          nboxes_to_generate, dev_sorted_scores.flat<float>().data(),
```

Comment:
We already have `dev_boxes_keep_flags` to tell which boxes to keep, why do we need to pass `dev_sorted_scores` and set scores of all not-selected boxes to min(float)?

Also from the code below we'll use `cub::DeviceSelect::Flagged` and `dev_boxes_keep_flags` to filter out those scores, so setting the score to min(float) becomes not useful?

Reply:
Hi @aaroey , that is correct. I removed score reference.

Reply:
Done